design-proposals/klip-52-bytes-data-type-support.md

```diff
@@ -0,0 +1,142 @@
+           
+## Motivation and background
+
+Currently, ksqlDB can only handle a set of primitive types and combinations of them.
+A BYTES data type would allow users to work with data that does not fit into any of
+the primitive types such as images, as well as BLOB/binary data from other databases.
+
+## What is in scope
+* Add BYTES type to KSQL
+* Support BYTE comparisons
+* Support BYTES usage in STRUCT, MAP and ARRAY
+* Serialization and de-serialization of BYTES to Avro, JSON, Protobuf and Delimited formats
+* Adding/updating UDFs to support the BYTES type
+
+## What is not in scope
+* Fixed sized BYTES (`BYTES(3)` representing 3 bytes, for example) - This is supported by Kafka Connect by adding the `connect.fixed.size`
+key in a bytes schema, but this will not be included in this KLIP.
+
+## Public APIS
+
+### BYTES
+
+The BYTES data type will store an array of raw bytes of an unspecified length. The maximum size of
+the array is limited by the maximum size of a Kafka message, as well as possibly by the value format being used.
+The syntax is as follows:
+
+```roomsql
+CREATE STREAM stream_name (b BYTES, COL2 STRING) AS ...
+CREATE TABLE table_name (col1 STRUCT<field BYTES>) AS ...
+```
+
+By default, BYTES will be displayed in console as HEX strings, where each byte is represented by two characters.
+For example, the byte array `[91, 67]` will be displayed as:
+
+```roomsql
+> SELECT b from STREAM;
+'0x5B43'
+```
+
+Implicit conversions to BYTES will not be supported.
+
+### UDF
+
+The following UDFs will be added:
+
+* `to_bytes(string, encoding)` - this will convert a STRING value in the specified encoding to BYTES.
+The accepted encoders are `hex`, `utf8`, `ascii`, and `base64`.
+* `from_bytes(bytes, encoding)` - this will convert a BYTES value to STRING in the specified encoding.
+The accepted encoders are `hex`, `utf8`, `ascii`, and `base64`.
+
+We will also update some of the existing STRING functions to accept BYTES as a parameter. In general, if a function works on ASCII characters for a STRING parameter,
+then it will work on bytes for a BYTES parameter.
+
+* `len(bytes)` - This will return the length of the stored ByteArray.
+* `concat(bytes...)` - Concatenate an arbitrary number of byte fields
+* `r/lpad(bytes, target_length, padding_bytes)` - pads input BYTES beginning from the left/right with the specified padding BYTES until the target length is reached.
+* `replace(bytes, old_bytes, new_bytes)` - returns the given BYTES value with all occurrences of `old_bytes` with `new_bytes`
+* `split(bytes, delimiter)` - splits a BYTES value into an array of BYTES based on a delimiter
+* `splittomap(bytes, entryDelimiter, kvDelimiter)` - splits a BYTES value into key-value pairs based on a delimiter and creates a MAP from them
+* `substring(bytes, to, from)` - returns the section of the BYTES from the byte at position `to` to `from`
```

Comment:
Substring with bytes feels awkward. I would expect the function to be called `slice`. Can we make the existing `slice` function that takes a collection argument work on bytes also?

Reply:
A lot of databases support `substring` for binary/blob, so that might actually be expected. 

In general, it seems like binary array types are seen as closer to text types than actual arrays.